spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.iceberg.ChangelogOperation;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterators;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+
+/**
+ * An iterator that transforms rows from changelog tables within a single Spark task. It assumes
+ * that rows are sorted by identifier columns and change type.
+ *
+ * <p>It removes the carry-over rows. Carry-over rows are unchanged rows in a snapshot but showed as
```

Comment:
I was working a little on re-wording this a bit? Feel free to use any of this if you like

Carry-over rows are the result of a removal and addition of the same row within an operation because of the copy-on-write mechanism. The iterator uses the "identifier fields" to determine of two rows are identical. For example, given a file which contains row1 (id=1, data='a') and row2 (id=2, data='b'). in a data file, a copy-on-write delete of row2 would require erasing this file. To preserve row1 a new file is written with row1' which is identical to row1. The change-log table would report this as (row1 deleted, row1' added), since this row was not actually modified it is not an actual change in the table.

Reply:
Adopted it in a new commit. I'm using `insert` instead of `add` to be consistent with our change type.